File: ios/App/App/Plugins/GoogleDrivePlugin/GoogleDrivePlugin.swift
import Foundation
import Capacitor
import GoogleSignIn
import GoogleAPIClientForREST_Drive

@objc(GoogleDrivePlugin)
public class GoogleDrivePlugin: CAPPlugin {

    private var driveService: GTLRDriveService?
    private var signInConfig: GIDConfiguration?

    override public func load() {
        super.load()

        // Configure Google Sign-In
        if let clientId = Bundle.main.object(forInfoDictionaryKey: "GOOGLE_CLIENT_ID") as? String {
            signInConfig = GIDConfiguration(clientID: clientId)
        }
    }

    private func checkConnection() throws {
        if driveService == nil {
            throw NSError(domain: "GoogleDrivePlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "unauthenticated"])
        }
    }

    @objc func connect(_ call: CAPPluginCall) {
        guard let signInConfig = signInConfig else {
            call.reject("Google Sign-In not configured. Missing GOOGLE_CLIENT_ID in Info.plist")
            return
        }

        let presentingViewController = bridge?.viewController

        GIDSignIn.sharedInstance.signIn(with: signInConfig, presenting: presentingViewController!) { user, error in
            if let error = error {
                call.reject("Google Sign-in failed", error.localizedDescription)
                return
            }

            guard let user = user else {
                call.reject("Google Sign-in failed", "No user returned")
                return
            }

            // Create Drive service
            self.driveService = GTLRDriveService()
            self.driveService?.authorizer = user.fetcherAuthorizer()

            call.resolve()
        }
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        GIDSignIn.sharedInstance.signOut()
        driveService = nil
        call.resolve()
    }

    @objc func store(_ call: CAPPluginCall) {
        do {
            try checkConnection()
        } catch {
            call.reject("Not authenticated", error.localizedDescription)
            return
        }

        guard let content = call.getString("content") else {
            call.reject("Content is required")
            return
        }

        // Create encrypted vault structure (simplified for now)
        let vault: [String: Any] = [
            "content": content,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: vault),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            call.reject("Failed to serialize vault data")
            return
        }

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        let filename = "\(dateFormatter.string(from: Date()))_encrypted_vault.json"

        // Create file metadata
        let file = GTLRDrive_File()
        file.name = filename
        file.mimeType = "application/json"
        file.parents = ["appDataFolder"]

        // Create file content
        let uploadParameters = GTLRUploadParameters(data: jsonString.data(using: .utf8)!, mimeType: "application/json")

        let query = GTLRDriveQuery_FilesCreate.query(withObject: file, uploadParameters: uploadParameters)
        query.fields = "id"

        driveService?.executeQuery(query) { ticket, file, error in
            if let error = error {
                call.reject("Failed to save to Google Drive", error.localizedDescription)
                return
            }

            call.resolve()
        }
    }

    @objc func fetchAllMetadata(_ call: CAPPluginCall) {
        do {
            try checkConnection()
        } catch {
            call.reject("Not authenticated", error.localizedDescription)
            return
        }

        let query = GTLRDriveQuery_FilesList.query()
        query.spaces = "appDataFolder"
        query.q = "mimeType='application/json' and trashed=false"
        query.fields = "files(id, name, createdTime)"
        query.orderBy = "createdTime desc"

        driveService?.executeQuery(query) { ticket, result, error in
            if let error = error {
                call.reject("Failed to fetch backup metadata", error.localizedDescription)
                return
            }

            guard let fileList = result as? GTLRDrive_FileList,
                  let files = fileList.files else {
                call.reject("Invalid response from Google Drive")
                return
            }

            let fileObjects = files.map { file -> [String: Any] in
                return [
                    "id": file.identifier ?? "",
                    "name": file.name ?? "",
                    "createdTime": file.createdTime?.dateTimeString ?? ""
                ]
            }

            call.resolve(["files": fileObjects])
        }
    }

    @objc func fetchFileContent(_ call: CAPPluginCall) {
        do {
            try checkConnection()
        } catch {
            call.reject("Not authenticated", error.localizedDescription)
            return
        }

        guard let fileId = call.getString("fileId") else {
            call.reject("fileId is required")
            return
        }

        let query = GTLRDriveQuery_FilesGet.queryForMedia(withFileId: fileId)

        driveService?.executeQuery(query) { ticket, file, error in
            if let error = error {
                call.reject("Failed to fetch backup content", error.localizedDescription)
                return
            }

            if let data = file as? Data,
               let content = String(data: data, encoding: .utf8) {
                call.resolve(["content": content])
            } else {
                call.reject("Invalid file content")
            }
        }
    }

    @objc func trash(_ call: CAPPluginCall) {
        do {
            try checkConnection()
        } catch {
            call.reject("Not authenticated", error.localizedDescription)
            return
        }

        guard let path = call.getString("path") else {
            call.reject("path is required")
            return
        }

        // First, find the file by name
        let listQuery = GTLRDriveQuery_FilesList.query()
        listQuery.spaces = "appDataFolder"
        listQuery.q = "name = '\(path)' and trashed = false"
        listQuery.fields = "files(id)"

        driveService?.executeQuery(listQuery) { ticket, result, error in
            if let error = error {
                call.reject("Failed to find backup file", error.localizedDescription)
                return
            }

            guard let fileList = result as? GTLRDrive_FileList,
                  let files = fileList.files,
                  let file = files.first,
                  let fileId = file.identifier else {
                call.reject("Backup file not found")
                return
            }

            // Now trash the file
            let updateFile = GTLRDrive_File()
            updateFile.trashed = true

            let updateQuery = GTLRDriveQuery_FilesUpdate.query(withObject: updateFile, fileId: fileId, uploadParameters: nil)

            self.driveService?.executeQuery(updateQuery) { ticket, file, error in
                if let error = error {
                    call.reject("Failed to delete backup", error.localizedDescription)
                    return
                }

                call.resolve()
            }
        }
    }
}

File: android/app/src/main/java/me/bitpoints/wallet/GoogleDrivePlugin.kt
package me.bitpoints.wallet

import android.app.Activity
import android.content.Intent
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.Scope
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import com.google.api.services.drive.model.File
import com.google.api.services.drive.model.FileList
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.util.*

@CapacitorPlugin(name = "GoogleDrive")
class GoogleDrivePlugin : Plugin() {

    companion object {
        private const val TAG = "GoogleDrivePlugin"
        private const val RC_SIGN_IN = 9001
    }

    private var driveService: Drive? = null
    private var pendingConnectCall: PluginCall? = null

    private val googleSignIn by lazy {
        GoogleSignIn.getClient(
            context,
            GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestScopes(Scope(DriveScopes.DRIVE_APPDATA))
                .build()
        )
    }

    private fun checkConnection() {
        if (driveService == null) {
            throw Exception("unauthenticated")
        }
    }

    private fun buildDriveService(account: GoogleSignInAccount) {
        val credential = GoogleAccountCredential.usingOAuth2(
            context,
            listOf(DriveScopes.DRIVE_APPDATA)
        ).apply {
            selectedAccount = account.account
        }

        driveService = Drive.Builder(
            NetHttpTransport(),
            GsonFactory.getDefaultInstance(),
            credential
        ).setApplicationName("Bitpoints Wallet").build()
    }

    private fun resolvePendingConnect() {
        val call = pendingConnectCall ?: return
        pendingConnectCall = null
        call.resolve()
    }

    private fun rejectPendingConnect(message: String, error: Throwable? = null) {
        val call = pendingConnectCall ?: return
        pendingConnectCall = null
        call.reject(message, error?.message, error)
    }

    override fun handleOnActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.handleOnActivityResult(requestCode, resultCode, data)

        if (requestCode != RC_SIGN_IN) {
            return
        }

        if (resultCode != Activity.RESULT_OK) {
            Log.e(TAG, "Google Sign-in cancelled or failed with resultCode=$resultCode")
            rejectPendingConnect("Google Sign-in cancelled")
            return
        }

        try {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            val account = task.result
            if (account == null) {
                Log.e(TAG, "Google Sign-in returned null account")
                rejectPendingConnect("Google Sign-in failed: account unavailable")
                return
            }

            CoroutineScope(Dispatchers.IO).launch {
                try {
                    buildDriveService(account)
                    withContext(Dispatchers.Main) {
                        resolvePendingConnect()
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to initialize Drive service", e)
                    withContext(Dispatchers.Main) {
                        rejectPendingConnect("Failed to initialize Google Drive", e)
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Google Sign-in error", e)
            rejectPendingConnect("Google Sign-in failed", e)
        }
    }

    @PluginMethod
    fun connect(call: PluginCall) {
        if (driveService != null) {
            call.resolve()
            return
        }

        val lastAccount = GoogleSignIn.getLastSignedInAccount(context)
        if (lastAccount != null) {
            CoroutineScope(Dispatchers.IO).launch {
                try {
                    buildDriveService(lastAccount)
                    withContext(Dispatchers.Main) {
                        call.resolve()
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to reuse last signed-in account", e)
                    withContext(Dispatchers.Main) {
                        call.reject("Failed to initialize Google Drive", e.message)
                    }
                }
            }
            return
        }

        synchronized(this) {
            if (pendingConnectCall != null) {
                call.reject("Google Sign-in already in progress")
                return
            }
            pendingConnectCall = call
        }

        try {
            val intent = googleSignIn.signInIntent
            startActivityForResult(call, intent, RC_SIGN_IN)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to launch Google Sign-in intent", e)
            rejectPendingConnect("Failed to launch Google Sign-in intent", e)
        }
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                googleSignIn.signOut()
                googleSignIn.revokeAccess()
                driveService = null
                withContext(Dispatchers.Main) {
                    call.resolve()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Google Sign-out error: $e")
                withContext(Dispatchers.Main) {
                    call.reject("Google Sign-out failed", e.message)
                }
            }
        }
    }

    @PluginMethod
    fun store(call: PluginCall) {
        checkConnection()

        val content = call.getString("content") ?: run {
            call.reject("Content is required")
            return
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val filename = "${SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())}_encrypted_vault.json"

                val fileMetadata = File().apply {
                    name = filename
                    mimeType = "application/json"
                    parents = listOf("appDataFolder")
                }

                val byteContent = content.toByteArray(Charsets.UTF_8)
                val mediaContent = com.google.api.client.http.ByteArrayContent("application/json", byteContent)

                val driveFile = driveService!!.files()
                    .create(fileMetadata, mediaContent)
                    .setFields("id")
                    .execute()

                Log.d(TAG, "File uploaded successfully: ${driveFile.id}")

                withContext(Dispatchers.Main) {
                    call.resolve()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to store backup: $e")
                withContext(Dispatchers.Main) {
                    call.reject("Failed to save to Google Drive", e.message)
                }
            }
        }
    }

    @PluginMethod
    fun fetchAllMetadata(call: PluginCall) {
        checkConnection()

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val result = driveService!!.files().list()
                    .setSpaces("appDataFolder")
                    .setQ("mimeType='application/json' and trashed=false")
                    .setFields("files(id, name, createdTime)")
                    .setOrderBy("createdTime desc")
                    .execute()

                val files = result.files.map { file ->
                    JSObject().apply {
                        put("id", file.id)
                        put("name", file.name)
                        put("createdTime", file.createdTime.toString())
                    }
                }

                val response = JSObject().apply {
                    put("files", files)
                }

                withContext(Dispatchers.Main) {
                    call.resolve(response)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to fetch metadata: $e")
                withContext(Dispatchers.Main) {
                    call.reject("Failed to fetch backup metadata", e.message)
                }
            }
        }
    }

    @PluginMethod
    fun fetchFileContent(call: PluginCall) {
        checkConnection()

        val fileId = call.getString("fileId") ?: run {
            call.reject("fileId is required")
            return
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val outputStream = ByteArrayOutputStream()
                driveService!!.files()
                    .get(fileId)
                    .executeMediaAndDownloadTo(outputStream)

                val content = outputStream.toString("UTF-8")

                val response = JSObject().apply {
                    put("content", content)
                }

                withContext(Dispatchers.Main) {
                    call.resolve(response)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to fetch file content: $e")
                withContext(Dispatchers.Main) {
                    call.reject("Failed to fetch backup content", e.message)
                }
            }
        }
    }

    @PluginMethod
    fun trash(call: PluginCall) {
        checkConnection()

        val path = call.getString("path") ?: run {
            call.reject("path is required")
            return
        }

        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Find file by name
                val fileList = driveService!!.files().list()
                    .setSpaces("appDataFolder")
                    .setQ("name = '$path' and trashed = false")
                    .setFields("files(id)")
                    .execute()

                val fileId = fileList.files?.firstOrNull()?.id
                    ?: throw Exception("Backup file not found")

                // Mark as trashed
                val file = File().setTrashed(true)
                driveService!!.files().update(fileId, file).execute()

                withContext(Dispatchers.Main) {
                    call.resolve()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to trash backup: $e")
                withContext(Dispatchers.Main) {
                    call.reject("Failed to delete backup", e.message)
                }
            }
        }
    }
}
